Add new NRT cache type DiscardReservedNodes

PiotrProkop opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind feature
<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:

Add new NRT cache type DiscardReservedNodes, which discards nodes from being scheduled if there are any pods reserved by NRT plugin for it.

DiscardReserved is intended to solve similiar problem as Overreserve Cache,
which is to minimize amount of incorrect scheduling decisions based on stale NRT data.
Unfortunately Overreserve cache only works for single-numa-node Topology Manager policy.
Dis tries to minimize amount of Admission Errors and non-optimal placement
when NodeResourceTopologyMatch plugin is used to schedule PODs requesting resources from multiple NUMA domains.
There are scenarios where using DiscardReserved won't mitigate drawbacks of using Passthrough cache.
NRT update is expected once PostBind triggers, but there's no guarantee about when this will happen.
In cases like:
- NFD(or any other component that advertises NRT) can be nonfunctional
- network can be slow
- Pod being scheduled after PostBind trigger and before NRT update
in those cases DiscardReserved cache will act same as Passthrough cache.


#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Add new DiscardReservedNodes cache type for NodeResourceTopologyPlugin.
```

PiotrProkop commented:
@ffromani @swatisehgal please review.

ffromani commented:
/ok-to-test
I had offline chat with @PiotrProkop and I'm very in favor of this general direction and open to this kind of additions. However I need to deep dive into the design and the code to see how we can clearly fit this code in the current framework, and unfortunately I won't have the time for that in the next few days.

ffromani commented:
@PiotrProkop just FYI additionally, I would finish https://github.com/kubernetes-sigs/scheduler-plugins/pull/488 before to tackle this one

PiotrProkop commented:
> @PiotrProkop just FYI additionally, I would finish #488 before to tackle this one

ok, let me know when #488 will be ready to review.

ffromani commented on pkg/noderesourcetopology/cache/overreserve.go:
```diff
@@ -70,27 +70,27 @@ func NewOverReserve(lister listerv1alpha1.NodeResourceTopologyLister, indexer No
 	nrt := ov.nrts.GetNRTCopyByNodeName(nodeName)
 	if nrt == nil {
-		return nil, true
+		return nil, fmt.Errorf("cannot get NodeResourceTopology for node: %s", nodeName)
```
This is a trap in the existing code and would have deserved a comment. For historical reasons, NOT having NRT info is a legit case which is meant to be handled as passthrough (no filter). We can revisit this behavior of course, but in another PR :)

ffromani replied:
I suggest to turn the error into a log and `return nil, nil` here

ffromani reviewed (commented):
still need to find time to properly review, but something caught my eye already

ffromani commented:
@PiotrProkop could you please rebase? I'd be happy to review the week starting on March 6 2023

PiotrProkop commented:
> @PiotrProkop could you please rebase? I'd be happy to review the week starting on March 6 2023

I am on it!

PiotrProkop commented:
> > @PiotrProkop could you please rebase? I'd be happy to review the week starting on March 6 2023
> 
> I am on it!

@ffromani  Should be good to review now!

ffromani commented:
> > > @PiotrProkop could you please rebase? I'd be happy to review the week starting on March 6 2023
> > 
> > 
> > I am on it!
> 
> @ffromani Should be good to review now!

Thanks! I'll start my review later today (March 6)

ffromani commented:
Sorry for the delay. Though week. Reviewing this PR is still on the top of my TODO.

ffromani commented on pkg/noderesourcetopology/cache/cache.go:
```diff
@@ -30,7 +30,7 @@ type Interface interface {
 	// Returns a boolean to signal the caller if the NRT data is clean. If false, then the node has foreign
 	// Pods detected - so it should be ignored or handled differently by the caller.
-	GetCachedNRTCopy(nodeName string, pod *corev1.Pod) (*topologyv1alpha2.NodeResourceTopology, bool)
+	GetCachedNRTCopy(nodeName string, pod *corev1.Pod) (*topologyv1alpha2.NodeResourceTopology, error)
```
this is probably a good idea in general. Ideally, this would be its own commit.

ffromani commented on pkg/noderesourcetopology/cache/cache.go:
```diff
@@ -56,4 +56,10 @@ type Interface interface {
+	// informational. A common application of this extension point is for cleaning
+	// up. If a plugin needs to clean-up its state after a pod is scheduled and
+	// bound, PostBind is the extension point that it should register.
+	PostBind(nodeName string, pod *corev1.Pod)
```
interesting. I'll review if this can helps the other cache types.

ffromani commented on pkg/noderesourcetopology/cache/passthrough.go:
```diff
@@ -34,17 +36,17 @@ func NewPassthrough(lister listerv1alpha2.NodeResourceTopologyLister) Interface
 	if err != nil {
-		klog.V(5).ErrorS(err, "Cannot get NodeTopologies from NodeResourceTopologyLister")
-		return nil, true
+		return nil, fmt.Errorf("cannot get NodeTopologies from NodeResourceTopologyLister")
```
this is another trap, the original code didn't fail. We need to check what happens on the call site with this change, which looks indeed more correct.

PiotrProkop replied:
this is what it looks like in lister:
```
	obj, exists, err := s.indexer.GetByKey(name)
	if err != nil {
		return nil, err
	}
	if !exists {
		return nil, errors.NewNotFound(v1alpha2.Resource("noderesourcetopology"), name)
	}
```
So if NRT doesn't exist we'll get Not Found error.

ffromani replied:
Uhm, I still tend to believe we DO NOT want to error out here, I recall some past conversations in this area. I'll do dome archeology.

ffromani replied:
let's hold on this and the other item regarding GetCachedNRTCopy. I'll do some research and come up with the suggestion. I'm confident postponing these changes should be little hassle.

ffromani commented on pkg/noderesourcetopology/filter.go:
```diff
@@ -198,12 +198,10 @@ func (tm *TopologyMatch) Filter(ctx context.Context, cycleState *framework.Cycle
-		return nil
+	nodeTopology, err := tm.nrtCache.GetCachedNRTCopy(nodeName, pod)
+	if err != nil {
+		klog.Error(err)
```
maybe `klog.ErrorS(err, ...)` and keep the original log message?

PiotrProkop replied:
:+1:

ffromani commented on pkg/noderesourcetopology/filter.go:
```diff
@@ -198,12 +198,10 @@ func (tm *TopologyMatch) Filter(ctx context.Context, cycleState *framework.Cycle
+	nodeTopology, err := tm.nrtCache.GetCachedNRTCopy(nodeName, pod)
+	if err != nil {
+		klog.Error(err)
+		return framework.NewStatus(framework.Unschedulable, err.Error())
```
why the other check was removed?

PiotrProkop replied:
In my first implementation I removed possibility of returning `nil,nil` from GetCachedNRTCopy, with next patchset I am going to revert it so will add this check again:
```go
if nodeTopology == nil {
	return nil
}
```

ffromani commented on pkg/noderesourcetopology/score.go:
```diff
@@ -63,15 +63,10 @@ func (tm *TopologyMatch) Score(ctx context.Context, state *framework.CycleState,
-	if nodeTopology == nil {
-		klog.V(5).InfoS("noderesourcetopology was not found for node", "node", nodeName)
-		return 0, nil
+		return framework.MinNodeScore, nil
```
again not sure (literally, need to check) that removing the `if nodeTopology == nil` guard has no ill effects in the flow.

PiotrProkop replied:
added this check again

ffromani replied:
thanks. Could you please add back the log as well?

PiotrProkop replied:
:+1:

ffromani commented on apis/config/v1/types.go:
```diff
@@ -149,6 +149,8 @@ type NodeResourceTopologyMatchArgs struct {
 	// If > 0, enables the caching facilities of the reserve plugin - which must be enabled
 	CacheResyncPeriodSeconds *int64 `json:"cacheResyncPeriodSeconds,omitempty"`
+	// if set to true, exclude node from scheduling if there are any reserved pods for given node
+	DiscardReservedNodes bool
```
missing JSON annotations. Please make sure the field is fully optional. Do you need to distinguish between "unspecified" and "zero"?

PiotrProkop replied:
I don't need to use `*bool` I can rely on default value of boolean which is `false` 😄

ffromani replied:
fair enough!

ffromani reviewed (commented):
initial partial review

PiotrProkop commented:
> initial partial review

thanks for the review. I have applied changes from this review.

ffromani commented:
For the record, overall from what I see so far I'm still very much convinced this change is a good direction for the NRT scheduler plugin and that the only issues could be related to properly integrate in the codebase - not about the design, which seems to fit nicely in the current framework and goals. There are already few points that could be generalized to other cache impls to make the codebase better.

PiotrProkop commented:
/retest

ffromani commented on pkg/noderesourcetopology/cache/discardreserved.go:
```diff
@@ -0,0 +1,98 @@
+		}
+	}
+
+	klog.V(5).InfoS("Lister for nodeResTopoPlugin", "lister", pt.lister)
```
minor: logging the lister is rarely useful in my experience. Perhaps you may want to log some properties of the lister, not the full object?

PiotrProkop replied:
You're right! I need to work on logging part of this PR.

ffromani commented on pkg/noderesourcetopology/cache/discardreserved.go:
```diff
@@ -0,0 +1,98 @@
+	if pt.reservationMap[nodeName] == nil {
+		pt.reservationMap[nodeName] = make(map[string]bool)
+	}
+	podName := klog.KObj(pod).String()
```
does this guarantee the name is stable and namespace prefixed?

PiotrProkop replied:
Haven't though about it, probably should use pod.UID.

PiotrProkop replied:
After checking I used this function cause it prefixes pod with namespace but I like using `UID` better, so I am changing implementation to use it right now!

ffromani commented on pkg/noderesourcetopology/cache/discardreserved.go:
```diff
@@ -0,0 +1,98 @@
+		pt.reservationMap[nodeName] = make(map[string]bool)
+	}
+	podName := klog.KObj(pod).String()
+	pt.reservationMap[nodeName][podName] = true
```
a design decision we inherited by the podresources api is to use the namespace/name pair to identify the pods.
Changing the podresources API would have been too hard, so this decision leaked all across the stack.
But pod.UID is a stronger identifier and in some case we would have preferred to use it.
So, is the namespace/name pair good enough for you? would using pod.UID lead to more robust/simpler code?

PiotrProkop replied:
pod.UID FTW!

ffromani commented on pkg/noderesourcetopology/cache/discardreserved.go:
```diff
@@ -0,0 +1,98 @@
+	podName := klog.KObj(pod).String()
+	pt.reservationMap[nodeName][podName] = true
+
+	klog.V(5).InfoS("nrtcache NRT", "logID", klog.KObj(pod), "node", nodeName, "reservationMap", pt.reservationMap[nodeName])
```
this easily gets big (thus poorly readable) when the nodes get populated, though. It's fine to log this only at high verbosiness level, but I'm wondering if we can log somehing more useful

ffromani commented on pkg/noderesourcetopology/cache/discardreserved.go:
```diff
@@ -0,0 +1,98 @@
+}
+
+// PostBind is invoked to cleanup reservationMap
+func (pt *DiscardReserved) PostBind(nodeName string, pod *corev1.Pod) {
```
this is interesting. I think I get it but could you please elaborate (gh comment is fine) why you want to remove the reservation on postbind?

ffromani replied:
note really looks very similar (possibly the same) reason we introduced the resync mechanism (using podfingerprint). You can't really be sure about timings here. The NRT update can come anytime. They will come in order, sure, but there's no guarantee about timing and about the sequence of events considering the scheduler event and the NRT update events.

ffromani replied:
Unless I'm missing something, you are totally correct that a NRT update is expected once `PostBind` triggers, but there's no guarantee about WHEN this will happen. NFD can be crashlooping, network can be slow, we can have any network glitch which prevents the NRT update to be propagated [temporary glitch]. But most notably there's no guarantee with respect the occurrence of new pods to be scheduled.

In yet another different way (but still related): what ensures that a NRT update is received before a new scheduling attempt is performed against this node when a new pod comes (new pod scheduling is of course completely async).

In the case of the overreserve, we have the podfingerprint machinery and we intentionally are using a pessimistic approach.

PiotrProkop replied:
You're right but `overreserve` doesn't cache anything if a POD requests resources from multiple NUMA domains(thanks to pesimistic approach and subtracting from every NUMA domain). And it will fetch data from NRT anyway (and temporary glitches also affects this 😄 ). The idea behind Locking nodes is to minimize Admission Errors not to eliminate them.

ffromani reviewed (commented):
almost done, need to finish looking at the tests and at the core changes

ffromani reviewed (commented):
Besides a couple items in the initial `GetCachedNRTCopy` PR, the prep work LGTM. I'll finish the review of the core changes later.

PiotrProkop commented:
> Besides a couple items in the initial `GetCachedNRTCopy` PR, the prep work LGTM. I'll finish the review of the core changes later.

Thanks a lot! I'll update PR till EOW.

PiotrProkop commented:
> > Besides a couple items in the initial `GetCachedNRTCopy` PR, the prep work LGTM. I'll finish the review of the core changes later.
> 
> Thanks a lot! I'll update PR till EOW.

PR updated. Decided to go with `pod.UID` as `reservationMap` key. I have also dropped the commit that changed the signature of `GetCachedNRTCopy`.

ffromani commented on apis/config/v1beta2/types.go:
```diff
@@ -148,6 +148,8 @@ type NodeResourceTopologyMatchArgs struct {
 	// If the cache is enabled, the Reserve plugin must be enabled.
 	CacheResyncPeriodSeconds *int64 `json:"cacheResyncPeriodSeconds,omitempty"`
+	// if set to true, exclude node from scheduling if there are any reserved pods for given node
+	DiscardReservedNodes bool `json:"discardReservedNodes,omitempty"`
```
my only remaining concern is this effectively makes the code ignore `CacheResyncPeriodSeconds` (if given).
I'd start documenting this. We can think about a better UX later on.

ffromani commented on pkg/noderesourcetopology/cache/discardreserved.go:
```diff
@@ -0,0 +1,91 @@
+func NewDiscardReserved(lister listerv1alpha2.NodeResourceTopologyLister) Interface {
+	return &DiscardReserved{
+		lister:         lister,
+		reservationMap: make(map[string]map[types.UID]bool),
```
minor: I wonder if we should use a set here (k8s.io/apimachinery/pkg/util/sets)

ffromani reviewed (commented):
thanks for the updates. Codewise LGTM with a note about the config, please check.
I'm OK adding this code to `noderesourcetopology` codebase, it's self-contained and I'm happy to assist your use case, but looks to me you folks are hitting very similar issues which led to the design of the `overreserve` cache, so maybe worth a deeper dive here?

ffromani commented:
(copypasting here to ensure visibility)

We covered all the facets I can think of. I'll do a last pass for due diligence to integration tests, but we are almost there. We need two things more, both should be quick and easy:

-  please document as comment in the config that enabling DiscardReservedNodes takes precedence over CacheResyncPeriodSeconds. A code comment is fine.
-  please describe (for historical record) the usecase and the design/conceptual differences between this cache impl and overreserve in the PR cover letter and/or in `cache/discardreserved.go` (preferably both, just copypaste the same block of text). 

I think we already discussed the rationale for this cache impl, it would be fine to just summarize our past conversation and copy/paste it as comment so we have the record for the future in a easy to find place.

PiotrProkop commented:
> (copypasting here to ensure visibility)
> 
> We covered all the facets I can think of. I'll do a last pass for due diligence to integration tests, but we are almost there. We need two things more, both should be quick and easy:
> 
>     * please document as comment in the config that enabling DiscardReservedNodes takes precedence over CacheResyncPeriodSeconds. A code comment is fine.
> 
>     * please describe (for historical record) the usecase and the design/conceptual differences between this cache impl and overreserve in the PR cover letter and/or in `cache/discardreserved.go` (preferably both, just copypaste the same block of text).
> 
> 
> I think we already discussed the rationale for this cache impl, it would be fine to just summarize our past conversation and copy/paste it as comment so we have the record for the future in a easy to find place.

Thanks for the review!

I have added comments in the code and updated PR description to explain rationale.

ffromani commented on pkg/noderesourcetopology/cache/discardreserved_test.go:
```diff
@@ -0,0 +1,172 @@
+			UID:       "some-uid",
+		},
+	})
+	nodePods, ok := nrtCache.reservationMap["node1"]
```
let's improve the tests later on
1. adding (more) integration tests
2. avoiding to access private fields/methods

ffromani reviewed (commented):
/approve
/lgtm

self-contained change. Useful to explore new directions which may benefit `overreserve in the future, so win-win
